### PR TITLE
General Hab2 Bug Fixes

### PR DESCRIPTION
**Pull request by ASzot**
## Motivation and Context

Bug fixes for Habitat 2.0 tasks:
- Spawning the agent on the largest navigation island.
- Bug fix for checking `on_top` predicate.
- Fixed bug with grasping manager not releasing grasp constraints.
- Refactored `hierarchical_policy`. 
- Compatibility with `rnn_build_seq_info=None` in distributed updater.

## How Has This Been Tested

End-to-end training.

## Types of changes

- **\[Refactoring\]** Large changes to the code that improve its functionality or performance
- **\[Bug Fix\]** (non-breaking change which fixes an issue)
- **\[Development\]** A pull request that add new features to the [habitat-lab](/habitat-lab) task and environment codebase. Development Pull Requests must be small (less that 500 lines of code change), have unit testing, very extensive documentation and examples. These are typically new tasks, environments, sensors, etc... The review process for these Pull Request is longer because these changes will be maintained by our core team of developers, so make sure your changes are easy to understand!

### CODE REVIEW

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_logical_expr.py [hunk 1]
```diff
@@ -108,6 +108,11 @@ def sub_in(self, sub_dict: Dict[PddlEntity, PddlEntity]) -> "LogicalExpr":
     def __repr__(self):
         return f"({self._expr_type}: {self._sub_exprs}"
 
+    @property
+    def compact_str(self):
+        sub_s = ",".join([s.compact_str for s in self._sub_exprs])
```
Comment (Skylion007) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_logical_expr.py:113]:
```suggestion
        sub_s = ",".join((s.compact_str for s in self._sub_exprs))
```

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_predicate.py [hunk 2]
```diff
@@ -106,8 +106,8 @@ def __repr__(self):
 
     @property
     def compact_str(self):
-        args = ",".join([str(x) for x in self._arg_values])
-        return f"{self._name}({args})"
+        args = ",".join([x.name for x in self._arg_values])
```
Comment (Skylion007) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_predicate.py:109]:
```suggestion
        args = ",".join((x.name for x in self._arg_values))
```

**Review by Skylion007 — APPROVED** [review]

File: habitat-baselines/habitat_baselines/rl/hrl/hierarchical_policy.py [hunk 3]
```diff
@@ -343,57 +385,72 @@ def act(
                     raise ValueError(
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/hrl/hierarchical_policy.py:385]: I think we should update hl_info here, as we did in the SIRo branch

**Review by xavierpuigf — COMMENTED** [review]
It looks generally correct, my only comment is on updating the hl_info properly when you call a skill, as we did in SIRo.
We should also update the hrl_rollout_storage with the bugs we saw right?

File: habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py [hunk 4]
```diff
@@ -23,16 +23,30 @@ def _recursive_apply(inp, fn):
         return fn(inp)
 
 
+def _convert_to_numpy_safe(t):
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py:26]: typing, is t supposed to be numpy array or torch tensor. 
If it is supposed to be a torch tensor, then it should not return t if it is on GPU

File: habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py [hunk 5]
```diff
@@ -23,16 +23,30 @@ def _recursive_apply(inp, fn):
...
-        inp, lambda t: t.numpy() if t.device.type == "cpu" else t
-    )
+    return _recursive_apply(inp, _convert_to_numpy_safe)
+
+
+def _numpy_to_cpu_safe(t):
+    if t is None:
+        return None
+    if isinstance(t, np.ndarray):
+        return torch.from_numpy(t)
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py:42]: I think if the default device has changed, this new tensor might not be on the CPU. You will need to explicitly put it there.

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py [hunk 6]
```diff
@@ -612,6 +612,9 @@ def expand_quantifiers(
         expanded_exprs: List[Union[LogicalExpr, Predicate]] = []
         assigns = []
         for poss_input in itertools.product(*all_matching_entities):
+            if len(set(poss_input)) != len(poss_input):
+                # Each of the `input` entries refers to a distinct entity.
+                continue
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py:617]: Shouldn't this raise an error ? If not, can you comment why ?

File: habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py (hunk 4 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py:26]: Addressed with docstring and typing.

File: habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py (hunk 5 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py:42]: Addressed with docstring and typing.

File: habitat-baselines/habitat_baselines/rl/hrl/hierarchical_policy.py (hunk 3 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/hrl/hierarchical_policy.py:418]: Good point, I included this.

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_logical_expr.py [hunk 7]
```diff
@@ -71,43 +80,65 @@ def is_true(self, sim_info: PddlSimInfo) -> bool:
         return self._is_true(lambda p: p.is_true(sim_info))
 
     def _is_true(self, is_true_fn) -> bool:
+        self._truth_vals = [None for _ in range(len(self._sub_exprs))]
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_logical_expr.py:83]:
```suggestion
        self._truth_vals = [None] * len(self._sub_exprs)
```

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_logical_expr.py [hunk 8]
```diff
@@ -71,43 +80,65 @@ def is_true(self, sim_info: PddlSimInfo) -> bool:
...
+                f"Got unexpected expr_type: {self._expr_type} of type {type(self._expr_type)}"
+            )
+
         if (
             self._expr_type == LogicalExprType.NAND
             or self._expr_type == LogicalExprType.NOR
         ):
-            ret = not ret
-        return ret
+            # Invert the entire result.
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_logical_expr.py:118]:
```suggestion
            # Invert the entire result for NAND and NOR expressions.
```

**Review by vincentpierre — APPROVED** [review]

File: habitat-baselines/habitat_baselines/common/construct_vector_env.py [hunk 9]
```diff
@@ -18,6 +18,7 @@ def construct_envs(
     config: "DictConfig",
     workers_ignore_signals: bool = False,
     enforce_scenes_greater_eq_environments: bool = False,
+    is_rank0: bool = True,
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/common/construct_vector_env.py:21]: Sure, I changed the name.

File: habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py [hunk 10]
```diff
@@ -23,22 +23,42 @@ def _recursive_apply(inp, fn):
         return fn(inp)
 
 
+def _convert_to_numpy_safe(t: Optional[torch.Tensor]) -> torch.Tensor:
+    """
+    Moves any tensors on the CPU to numpy arrays. Leaves other values unaffected.
+    """
+
+    if t is not None and t.device.type == "cpu":
+        return t.numpy()
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/ddppo/algo/ddppo.py:32]: No, it's important we don't have it to catch problems.

File: habitat-baselines/habitat_baselines/rl/hrl/hierarchical_policy.py [hunk 11]
```diff
@@ -94,6 +101,15 @@ def __init__(
         self._stop_action_idx, _ = find_action_range(
             action_space, "rearrange_stop"
         )
+        first_idx: Optional[int] = None
+        for skill_i, skill in self._skills.items():
+            if self._idx_to_name[skill_i] == "noop":
+                continue
+            if isinstance(skill, NoopSkillPolicy):
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/hrl/hierarchical_policy.py:108]: Fixed this logic.

**Pull request merged**